### PR TITLE
Upgrade to Ink 3

### PR DESCRIPTION
**Pull request by vadimdemedes**
CleanShot 2020-06-23 at 21 16 53

Also rewritten the whole thing using hooks.

### CODE REVIEW

**Comment by sindresorhus** [comment]
After searching, if I press Escape to leave, I get this:

```
❯ node cli.js

› horse



Warning: Can't perform a React state update on an unmounted component. This is a no-op, but it indicates a memory leak in your application. To fix, cancel all subscriptions and asynchronous tasks in the componentWillUnmount method.
    in Emoj (created by Context.Consumer)
    in Unknown
    in App
```

**Comment by sindresorhus** [comment]
Thanks for upgrading this. The changes look great and much more readable now with hooks.

File: ui.js [hunk 1]
```diff
@@ -21,47 +21,32 @@ const STAGE_CHECKING = 0;
...
+	</Text>
 );
 
 const Search = ({query, emojis, skinNumber, selectedIndex, onChangeQuery}) => {
 	const list = emojis.map((emoji, index) => (
-		<Color
+		<Text
 			key={emoji}
-			bgGray={index === selectedIndex}
+			backgroundColor={index === selectedIndex ? 'gray' : undefined}
```
Comment (sindresorhus) [inline ui.js:44]:
```suggestion
			backgroundColor={index === selectedIndex && 'gray'}
```

File: ui.js [hunk 2]
```diff
@@ -78,80 +63,59 @@ const Search = ({query, emojis, skinNumber, selectedIndex, onChangeQuery}) => {
...
-		);
-	}
+const Emoj = ({skinNumber: initialSkinNumber, onSelectEmoji}) => {
+	const {exit} = useApp();
+	const [stage, setStage] = useState(STAGE_CHECKING);
+	const [query, setQuery] = useState('');
+	const [emojis, setEmojis] = useState([]);
+	const [skinNumber, setSkinNumber] = useState(initialSkinNumber);
+	const [selectedIndex, setSelectedIndex] = useState(0);
+	const [selectedEmoji, setSelectedEmoji] = useState(null);
```
Comment (sindresorhus) [inline ui.js:73]: Is it possible to use `undefined` here instead of `null`? I'm trying to abolish `null` from my projects.

File: ui.js [hunk 3]
```diff
@@ -78,80 +63,59 @@ const Search = ({query, emojis, skinNumber, selectedIndex, onChangeQuery}) => {
...
+	useEffect(() => {
+		setStage(STAGE_SEARCH);
+	}, []);
+
+	useEffect(() => {
+		if (query.length <= 1) {
+			return;
+		}
+
+		debouncer(async () => {
```
Comment (sindresorhus) [inline ui.js:96]: Any debouncer hook we could use? Might make this slightly nicer.

Alternatively, maybe `useInput` could have a `debounce` option? Seems like a common use-case to want to debounce input.

Comment (vadimdemedes) [reply ui.js:96]: > Any debouncer hook we could use? Might make this slightly nicer.

Good point.

> Alternatively, maybe useInput could have a debounce option? Seems like a common use-case to want to debounce input.

Could be worth adding, going to wait and see if this comes up again.

**Comment by vadimdemedes** [comment]
Fixed everything, including that warning on exit:

```
Warning: Can't perform a React state update on an unmounted component. This is a no-op, but it indicates a memory leak in your application. To fix, cancel all subscriptions and asynchronous tasks in the componentWillUnmount method.
    in Emoj (created by Context.Consumer)
    in Unknown
    in App
```

**Pull request merged**